Proposals/20200511-HVPA-design.md

```diff
@@ -0,0 +1,129 @@
+# HVPA-EP-0001: HVPA Redesign
+
+## Summary
+This proposal outlines a plan to enable HVPA to do more efficient scaling horizontally and vertically.
+
+## Motivation
+Currently, HVPA scales by applying user provided weights to HPA and VPA recommendations. However, if 100% weight is provided to VPA, then horizontal scaling will never happen.
```

Comment:
> about the memory cache

This approach does not address the cache problem, because the VPA recommendations themselves are based on current memory usage, which has duplicate caches accounted for anyway.

> the ceiling problem

We did start from here, however, we have ended up there again, i believe for all the right reasons.
Following snippet is from the discussion we had over email:
```
Current Requests: 150m / 1.8Gi with 1 replica
VPA Recommendation: 150m / 2.2Gi
New Requests: 150m / 2.2Gi with 1->2 replicas, because 2.2Gi lands us in the “second” bucket when out-scaling that has 2 replicas; we carry over the vertical scaling as is, which is most conservative.
```
In the old weight based approach, if the weight of 50% was provided, we would have ended up with:
2 replicas of `150m / 2GB` each

These points are discussed in further detail under sub-section `Base usage of each replica` in this document.

Reply:
Just to understand the difference between the approaches:
**In old approach:**
we cap/ceil the replicas, and also have a partial vertical scaling. This results in overscaling especially if HPA recommends in increments of 1. If I understand right, this is the issue mentioned in your comment.

**In new approach:**
Whenever we scale up resources within a bucket, we cap/ceil cpu/memory as mentioned under section `Base usage of each replica`. 

Haven't we just shifted the vagaries from replicas to cpu/memory in the new approach?
To mitigate the overscaling in the new approach we have proposed to tune VPA so that it reacts quicker to overscaling. This would be possible only for VPA releases 0.8.0+ when [these](https://github.com/kubernetes/autoscaler/pull/2905) changes are released.
However, same or similar mitigation can also be proposed for older approach. In fact, HPA tuning is possible to do already, using HPA scale down delay.

This is the reason why in my previous comment I mentioned:
> We did start from here, however, we have ended up there again

Because of this reason I am reluctant in documenting this under "motivation" of taking this approach because new approach doesn't change the core nature of the problem, it just shifts the focus.
Would it be ok if I add these details under pros and cons of each of the approaches/alternatives?

Reply:
No. My point was not overscaling. That cannot be avoided, in principle. 

My point was the _uncontrolled_ nature of scaling. The old approach of weightage leads to scaling behaviour (amount of vertical scaling vs horizontal scaling) that is subject to the vagaries of recommendation from HPA. If it recommends in increments of 1 then the behaviour is different and if it recommends in larger increments then the behaviour is different. Then there was also the problem of what to use to determine the buckets. The current replicas? recommended replicas? Renormalize by a recursion of choosing a bucket, adjusting the scaling and then choosing a bucket according to the adjusted scaling?

The new approach of using the _abosolute_ scale to determine buckets was fundamentally to address these problems. Not overprovisioning.

Reply:
> My point was the uncontrolled nature of scaling
ah i see. Will add this to doc now. Sorry, i don't remember discussing this as a reason/motivation.

> Then there was also the problem of what to use to determine the buckets. The current replicas? recommended replicas? Renormalize by a recursion of choosing a bucket, adjusting the scaling and then choosing a bucket according to the adjusted scaling?

As far as this is concerned, there should not be any confusion. The process of choosing a bucket is defined unambiguously under `alternative approach`
We decide based on **current** replicas, unless upper/lower thresholds for cpu/memory are crossed by VPA recommendations. If thresholds are crossed, we take next/previous bucket's weight as described in `Alternative approach` section. Also, there is **no normalisation**, or adjustment based on "chosen bucket". Only the `vpaWeight` of the chosen bucket is considered.

Reply:
Regarding the clarity in the alternative approach, yes. My point here was what current replicas or any other way to decide the bucket in the alternative approach cannot lead to a controlled way of scaling horizontally and vertically. We have discussed already that any such approach that doesn't take the _absolute_ scale into consideration will not lead to a controlled way of scaling. That was the other motivation to go for the proposal based on _absolute_ scale. Not that you have not mentioned the current replicas clearly in the alternative approach.

Reply:
Considering points mentioned under section `Base usage of each replica`, the final scaling is anyway not according to *absolute scale*. The *absolute scale* is now only used to get to the right bucket/replicas. After finalising the resource size according to max value, the actual aggregate desired resources may or may not fall in the chosen bucket.
Would this scaling still be called "controlled"?

Reply:
Yes. Certainly more than the alternative.

Also, I meant about the absolute scale being used for deciding the buckets. Nothing more.

Reply:
👍 Updated the section accordingly.